File: Sorting_algorithms/Kotlin/merge_sort.kt
fun arr_merge_agent(left_arr: IntArray, right_arr: IntArray): IntArray {
    var ord_arr = mutableListOf<Int>()
    var left_idx = 0
    var right_idx = 0

    while (left_idx < left_arr.size && right_idx < right_arr.size) {
        if (left_arr[left_idx] <= right_arr[right_idx]) {
            ord_arr.add(left_arr[left_idx])
            left_idx++
        } else {
            ord_arr.add(right_arr[right_idx])
            right_idx++
        }
    }
    ord_arr.addAll(left_arr.toList().toMutableList().subList(left_idx, left_arr.size))
    ord_arr.addAll(right_arr.toList().toMutableList().subList(right_idx, left_arr.size))
    return ord_arr.toList().toIntArray()
}


fun merge_sort (arr: IntArray): IntArray {
    if (arr.size <= 1 ) {
        return arr
    } else {
        val cut_point = arr.size / 2
        var left = merge_sort(arr.copyOfRange(0, cut_point))
        var right = merge_sort(arr.copyOfRange(cut_point, arr.size))
        return arr_merge_agent(left, right)
    }
}


fun main () {
    var arrList = intArrayOf(11, 30, -1, 2, 7, 12, 13, 20, 8)
    println("\nBefore Sort: ${arrList.joinToString(", ")}")
    arrList = merge_sort(arrList)
    println("After Sort: ${arrList.joinToString(", ")}")
}

File: Sorting_algorithms/python/src/merge_sort .py
def arr_merge_agent(left_arr:list, right_arr:list):
    ord_arr = []
    left_idx, right_idx = 0,0
    while left_idx < len(left_arr) and right_idx < len(right_arr):
        if left_arr[left_idx] <= right_arr[right_idx]:
            ord_arr.append(left_arr[left_idx])
            left_idx+=1
        else:
            ord_arr.append(right_arr[right_idx])
            right_idx+=1
    ord_arr += left_arr[left_idx:]
    ord_arr += right_arr[right_idx:]
    return ord_arr


def merge_sort(arr:list):
    if len(arr) <= 1: return arr
    else:
        cut_point = len(arr)//2
        left = merge_sort(arr[:cut_point])
        right = merge_sort(arr[cut_point:])
        return arr_merge_agent(left, right)


if __name__ == "__main__":
    arr:list = [11, 30, -1, 2, 7, 12, 13, 20, 8]
    print(f"\nBefore Sort: {arr}")
    print(f"After Sort: {merge_sort(arr=arr)}\n")
